### ISSUE TITLE
Using navigationStart as a base line may expose cross-origin timing information

### ISSUE THREAD

**Issue by noamr** [w3c/navigation-timing#160]
When we have a navigation with cross-origin redirects, we're hiding `redirectStart` and `redirectEnd` from the final document.

However, because the base timestamp for all the navigation timing entries is the navigation start, the redirect timing info can (somewhat) easily be inferred.

Consider the following:

- User clicks on URL from domain A at timestamp ts1 (e.g. a search engine click handler URL or an ad broker like outbrain)
- Domain A takes a while to handle the request
- The request redirects to domain B at timestamp ts2
- Domain B handles the request and serves the document
- ts1 is available to the document, directly or indirectly, as it's the `navigationStart` which is the base timestamp for all navigation timing / resource timing entries.

I believe we have three ways to go about it (but maybe there are more):
- Enable all the navigation timing properties, based on the notion that the cross-origin information is already exposed by navigationStart.
- Change `navigationStart` to be the first redirect in the current origin redirect chain
- Make use of `TAO` (in its current form or with some amendments) give redirect chains to opportunity to expose their timing to the destination.

This came from discussing whether to enable or zero-out navigation timing properties.
See previous discussions [here](https://github.com/whatwg/html/pull/7105), [here](https://github.com/web-platform-tests/wpt/pull/30978) and [here](https://github.com/w3c/navigation-timing/issues/117).

Thoughts?

**Comment by noamr** [comment]
@yoavweiss @annevk @npm1 @sefeng211 @achristensen07

**Comment by yoavweiss** [comment]
Also @bdekoz @mikewest @arturjanc

**Comment by annevk** [comment]
If "current origin redirect chain" means the origin boundary is not crossed therein, that option seems reasonable to me. The current model does indeed seem problematic.

https://github.com/w3c/resource-timing/issues/220 is also related.

**Comment by noamr** [comment]
> If "current origin redirect chain" means the origin boundary is not crossed therein, that option seems reasonable to me. 

Yes, it means something like "after all the redirects that are not same-origin as the document's final origin are complete"

**Comment by achristensen07** [comment]
This does need fixing.  I oppose the first option.  The third option seems most like what we do with other cross-origin timing exposure.

**Comment by mikewest** [comment]
I think I agree with @annevk and @achristensen07. This does seem like something we ought to change, and the third option seems like the most robust (and consistent) way of doing so. I think the second is justifiable from a security standpoint as well, but I'm not sure the complexity it introduces is worthwhile.

**Comment by noamr** [comment]
Note that (3) includes (2) inside it, in the cases where TAO headers are not there.

With option 3 I want to be careful when we overload the meaning of TAO (see concerns [here](https://github.com/w3c/resource-timing/issues/223). Though it could be that since TAO was ignored inside redirects so far it's not a problem since it's new usage of an existing header.

Perhaps a good way to go about it would be that the TAO Header would have to specify the same origin as the `Location` header, rather than `*`, like this:

`
Location: my-site.com/page.html
Timing-Allow-Origin: my-site.com
`

Not having the above in a cross-origin redirect would push the navigation start time (and `timeOrigin`s) to the beginning of the next fetch after this redirect.

**Comment by annevk** [comment]
As I noted in https://github.com/w3c/resource-timing/issues/220 accounting for TAO here is in essence a new model and therefore also a source of complexity. For Resource Timing we have a document A that fetches B which redirects to C. B and C need to consent. Here we have a document A that is navigated to B which redirects to C. B needs to consent? I think it's acceptable, but it's quite a bit different.

**Comment by yoavweiss** [comment]
I have to say that the second and third options are not mutually exclusive - we could set navigation time at the first same-origin redirect and extend it backwards pending on an opt-in from the cross-origin redirects.

Note that in any case, we'd need *some* point in time for navigation start even if there's no opt-in. At worst, we can have that be the request to the eventual request for the document, but setting it to the first request in the last same-origin portion of the redirect chain doesn't seem overly complex.

**Comment by yoavweiss** [comment]
> As I noted in [w3c/resource-timing#220](https://github.com/w3c/resource-timing/issues/220) accounting for TAO here is in essence a new model and therefore also a source of complexity. For Resource Timing we have a document A that fetches B which redirects to C. B and C need to consent. Here we have a document A that is navigated to B which redirects to C. B needs to consent? I think it's acceptable, but it's quite a bit different.

I agree that TAO may not be the opt-in we want here.

**Comment by noamr** [comment]
> I agree that TAO may not be the opt-in we want here.

Opting for a new opt-in? 😬

**Comment by jyasskin** [comment]
Yoav pointed out that in cases where the redirection domain is trying to help the source and destination track the user, and the user agent has blocked the query parameters they had been using to do this, this timing information might help them continue to transfer at least a partial identifier. I don't think the navigational-tracking threat model (https://github.com/privacycg/nav-tracking-mitigations/issues/12) is developed enough to be much help in making decisions here, but heads up that the Privacy CG might come back to this later.

**Comment by noamr** [comment]
> Yoav pointed out that in cases where the redirection domain is trying to help the source and destination track the user, and the user agent has blocked the query parameters they had been using to do this, this timing information might help them continue to transfer at least a partial identifier. I don't think the navigational-tracking threat model ([privacycg/nav-tracking-mitigations#12](https://github.com/privacycg/nav-tracking-mitigations/issues/12)) is developed enough to be much help in making decisions here, but heads up that the Privacy CG might come back to this later.

Hmm this makes it more interesting - it means that if we accept this as a threat to be mitigated, the user agent should have a say in this for the purpose of tracking prevention and not just the two domains, which means something along option (2) would be the (only?) way to go (the "navigation" starts from the last same-origin chain).

**Comment by yoavweiss** [comment]
Indeed. As long as query parameters are allowed to be passed with the navigation URLs, acting against this doesn't matter much. So for now, I don't think we should take that into account.

But if and when we start mitigating query parameters as an information-passing channel, we'd need to also mitigate the redirection timing channel, either by not exposing it entirely, or have browsers lie about those times in smarter ways (e.g. for known trackers, when they're highly variable, etc).

**Comment by annevk** [comment]
There are user agents doing experiments around query parameters (and some might have shipped?) so we might as well account for it now.

**Comment by yoavweiss** [comment]
I agree this is an area we should remain vigilant on to see how it develops, but I don't believe we have consensus on the threat model and what the solutions to this threat would look like. So it seems premature to e.g. eliminate an opt-in option before that settles.

**Comment by annevk** [comment]
Maybe, I have to say that since we have some tentative plans in this area, I'm actually hesitant now to support an opt-in model here.

**Comment by yoavweiss** [comment]
^^ @miketaylr

It's true that we could go with option (2) and expand it later with an opt-in, once things in that area settle.

It may also be interesting to think about the incentive model here - if this would make redirectors unaccountable for their performance, that'd not be a great outcome. An opt-in model may not be effective in driving such accountibility.

**Comment by noamr** [comment]
> ^^ @miketaylr
> 
> It's true that we could go with option (2) and expand it later with an opt-in, once things in that area settle.

Feels to me from this discussion that this might be the way forward as a first step, while contemplating the opt-in. Interesting if something else would come up in TPAC.

> It may also be interesting to think about the incentive model here - if this would make redirectors unaccountable for their performance, that'd not be a great outcome. An opt-in model may not be effective in driving such accountibility.

I'm not sure a redirecting URL is accountable for timing to its destination... Maybe it's accountable to the domain that started the navigation, e.g. where the banner was?

Maybe the interested party in this information is neither domain, but rather the user (and the user agent), and user agents should be encouraged to show some UI indication during a cross-origin navigation redirect, to show the user that the delay comes from an ad broker etc and not the originating domain / destination domain, rather than counting on the origin/destination URLs to do something about it? </Thoughts.>

**Comment by npm1** [comment]
Just to clarify, sounds like this applies to https://w3c.github.io/hr-time/#dfn-time-origin as well? If it does then this would be a pretty big change impacting any high resolution timestamps received by developers.

**Comment by yoavweiss** [comment]
> Just to clarify, sounds like this applies to https://w3c.github.io/hr-time/#dfn-time-origin as well? If it does then this would be a pretty big change impacting any high resolution timestamps received by developers.

Indeed! We'll definitely have to be careful about rolling this out.

**Comment by noamr** [comment]
See demo [here](https://github.com/web-platform-tests/wpt/pull/31086). A minimal use case without an intermediate domain. 

The originating domain does some form processing of `POST` before redirecting to a URL at a different domain.
Because the `timeOrigin` includes the time it took to process the `POST` at the originating domain, in this case the destination domain can detect whether the user subscribed to a newsletter.

**Comment by noamr** [comment]
Thank you @sgomes, yes, this voice has to be heard to.
Seems like the implications of this on RUM metrics would be pretty big. That would have to be weighed against the current breach of same-origin policy. Hoping for a lively discussion at TPAC.

**Comment by annevk** [comment]
I think the bigger problem is that when navigating from A to B, B gets to learn when the action on A happened, in quite some precision.

While currently there are many side channels through which A and B can keep tabs on the user, at some point this timing channel might be all that's left and it seems rather unfortunate that it's there.

**Comment by terjanq** [comment]
@noamr yes I agree, that it's not an attack. The word attack in my response was to convince myself that it's not a side channel that could be used, but rather potential privacy leak about past. Speaking of past, a website B could potentially infer the information: "There was at least one unexpected redirect and it took X time", but it doesn't know exactly about other origins? It knows that something like this occurred, but it doesn't give more information at first glance? 

From the adversary point of view, I could argue that the time it takes for a page to load leaks, with a high likelihood, what redirector was used to load a page. (e.g. facebook redirectors could take 79ms, while twitter 10ms), which then leaks information from what origin the user visited the page, even if document.referrer was supposed to be hidden. That indeed might seem problematic.

**Comment by noamr** [comment]
In the meantime, added a PR to [move time origin computation to HTML](https://github.com/whatwg/html/pull/7339), this would hopefully make it easier to see how it works and make changes in this in the future.

**Comment by andydavies** [comment]
Metrics such as LCP are trying to represent the user's experience and if navgiationStart is changed to be before the first redirect for the current origin then it will no-longer represent the user's experience for many cases

Visitors clicking from search engines / social media sites are typically routed through redirect, in the worst case clicking on a Ad in Google results leads to three redirects that consume nearly a second for example

The time before the first redirect on the current origin is important as it reflects the visitors actual experience

**Comment by noamr** [comment]
> Metrics such as LCP are trying to represent the user's experience and if navgiationStart is changed to be before the first redirect for the current origin then it will no-longer represent the user's experience for many cases
> 
> 
> 
> Visitors clicking from search engines / social media sites are typically routed through redirect, in the worst case clicking on a Ad in Google results leads to three redirects that consume nearly a second for example
> 
> 
> 
> The time before the first redirect on the current origin is important as it reflects the visitors actual experience
> 
> 

Understood, but nonetheless the cross-domain leak is there, and the fact that removing it would hinder this valuable metric doesn't make it less of a leak.

**Comment by yoavweiss** [comment]
@andydavies - I think the use case for having visibility to these cross-origin redirects is clear. At the same time, this is a cross-origin leak, as it is exposed right now. We'd need to find an alternative, privacy-preserving way to expose this information to developers. (e.g. aggregated reporting)

**Comment by noamr** [comment]
Suggesting to start tackling this by adding an opt-in.
Something similar to TAO but specific to Redirect to avoid loading new meanings on old headers.

How about: `Redirect-Timing-Allow-Origin: same-origin | same-site | [url list] | *` (default `same-origin`)
Not sure if we want it closer to TAO or to only allow same-origin/same-site.

At first, it would allow exposing redirectStart/redirectEnd when there are cross-origin redirects in navigation, and later on we can make this as the only way to expose cross-origin redirects at start-of-navigation.

**Comment by noamr** [comment]
> I don't see how an opt-in header helps with the attacks? See also [#160 (comment)](https://github.com/w3c/navigation-timing/issues/160#issuecomment-952000354).

You're right, it doesn't help with that aspect.

**Comment by yoavweiss** [comment]
https://github.com/WebKit/WebKit/pull/63781#issuecomment-4353113185 changes how I thought about this, as apparently WebKit doesn't change the timeOrigin to the last redirect on the destination origin, but instead nullifies most attributes when TAO is not present across the entire redirect chain.

I think we have a few options:
1) Do nothing and live with significant interop issues (e.g. no TTFB in WebKit if there's a cross-origin redirect without TAO) 
2) Move the `timeOrigin` to the final navigation, and add a `redirectDuration` attribute. The latter could require an opt-in, at least in some implementations.
  - We can probably drop `redirectStart` and `redirectEnd` 
3) Similar to (2), but moving the `timeOrigin` to the first request in the redirect chain on the destination origin.
4) ???

I think (1) results in both confusion (developers not knowing that their metrics are influenced by redirects) and interop (missing attributes). (2) seems to me less confusing than (1).

**Comment by mmocny** [comment]
It sounds like the results of https://github.com/WebKit/WebKit/pull/63781#issuecomment-4353113185 are that the current situation is *more* interoperable than we previously thought.

There remain gaps (such as missing redirect values), but those were already there previously, and at least the current `timeOrigin` value indirectly gives some insights (via ["UNO"](https://calendar.perfplanet.com/2024/uno/))

**Comment by noamr** [comment]
(4) expose responseStart, but hide everything between startTime and responseStart when there are cross origin redirects without TAO.

responseStart is already observable by other means, like scripts or link headers.

This would keep TTFB as well as LCP etc interoperable, while not exposing how much time is spent in cross origin redirects in a way that's easily discernable from same-origin request preparation, DNS,  connection etc.

For webkit this would means keeping the current implementation but always unhiding responseStart for navigation timing.

**Comment by yoavweiss** [comment]
Thinking about this again.. I suspect that (4) is the fastest (only?) path to interop here.

A few things we need to figure out:
1) Can RUM providers live with the lack of timestamps up to `responseStart` in case of an no-TAO cross-origin redirect? /cc @nicjansma @krzksz @andydavies
2) Can Safari live with exposing responseStart in case of no-TAO cross-origin redirect, and exposing all the timestamps in case of TAO? /cc @achristensen07 
3) Can Chrome/Firefox change their behavior to exclude timestamps up to `responseStart` for no-TAO cross-origin redirects? /cc @noamr @mmocny @Bas-moz 

One more thing to think about - For the redirect timestamps, we also [excluded them in the case of a `no-referer` request](https://html.spec.whatwg.org/#:~:text=navigationParams%27s%20request%27s%20client%20is%20null%2C%20or%20navigationParams%27s%20request%27s%20referrer%20is%20not%20%22no%2Dreferrer%22%2C%20and). Should we do the same for the rest of the attributes here?

**Comment by noamr** [comment]
> Thinking about this again.. I suspect that (4) is the fastest (only?) path to interop here.
> 
> A few things we need to figure out:
> 
> 1. Can RUM providers live with the lack of timestamps up to `responseStart` in case of an no-TAO cross-origin redirect? /cc [@nicjansma](https://github.com/nicjansma) [@krzksz](https://github.com/krzksz) [@andydavies](https://github.com/andydavies)
> 2. Can Safari live with exposing responseStart in case of no-TAO cross-origin redirect, and exposing all the timestamps in case of TAO? /cc [@achristensen07](https://github.com/achristensen07)
> 3. Can Chrome/Firefox change their behavior to exclude timestamps up to `responseStart` for no-TAO cross-origin redirects? /cc [@noamr](https://github.com/noamr) [@mmocny](https://github.com/mmocny) [@Bas-moz](https://github.com/Bas-moz)

I think I am OK with the behavior but cannot commit on a timeline for doing the work. Are you taking it on @yoavweiss ?

> One more thing to think about - For the redirect timestamps, we also [excluded them in the case of a `no-referer` request](https://html.spec.whatwg.org/#:~:text=navigationParams%27s%20request%27s%20client%20is%20null%2C%20or%20navigationParams%27s%20request%27s%20referrer%20is%20not%20%22no%2Dreferrer%22%2C%20and). Should we do the same for the rest of the attributes here?

**Comment by yoavweiss** [comment]
> I think I am OK with the behavior but cannot commit on a timeline for doing the work. Are you taking it on [@yoavweiss](https://github.com/yoavweiss) ?

Not a hard commitment, but assuming we all agree on an interoperable path, I could most probably take on the Chromium/WebKit work.
